### ISSUE TITLE
Blockers for v1.2.0

### ISSUE THREAD

**Issue by AkihiroSuda** [opencontainers/runc#4114]
Blockers for rc.1 (https://github.com/opencontainers/runc/pull/3963):
- [x] https://github.com/opencontainers/runc/pull/4101
- [x] https://github.com/opencontainers/runc/pull/3985
- [x] https://github.com/opencontainers/runc/pull/3990
- [x] https://github.com/opencontainers/runc/issues/4125
- [x] https://github.com/opencontainers/runc/issues/4172
- [x] https://github.com/opencontainers/runc/issues/4173

Blockers for GA:
- [x] https://github.com/opencontainers/runc/issues/4093
- [x] Release runtime-spec ~v1.1.1~ v1.2.0 https://github.com/opencontainers/runtime-spec/pull/1242
- [x] https://github.com/opencontainers/runc/issues/4233
- [x] https://github.com/opencontainers/runc/pull/4283#issuecomment-2156033671
- [x] Verify compatibility with well-known downstreams
  - [x] https://github.com/moby/moby/pull/47666
    - [x] Fix failing moby tests in CI https://github.com/opencontainers/runc/issues/4384
  - [x] https://github.com/containerd/nerdctl/pull/3153
    - [x]  https://github.com/opencontainers/runc/issues/4328
    - [x] https://github.com/opencontainers/runc/issues/4394
  - [x] containerd works fine, CI scripts need fixes
    - [x] PR to fix containerd's CI scripts: https://github.com/containerd/containerd/pull/10449
    - [x] cri-tools PR backported for v1.30.1: https://github.com/kubernetes-sigs/cri-tools/pull/1492
- [x] https://github.com/opencontainers/runc/issues/4294

Needs discussion (probably deferrable to v1.2.1 or v1.3.0):
- [ ] https://github.com/opencontainers/runc/issues/3922

@opencontainers/runc-maintainers Feel free to edit this issue

**Comment by AkihiroSuda** [comment]
We don't have timeline for v1.2.0, but we should try to release v1.2.0 by the end of the year.

**Comment by kolyshkin** [comment]
Well, we have the 1.2.0 milestone (which I am trying comb through regularly), and everything that's in there is kind of a blocker for 1.2.0.

Feel free to modify milestones for any issues/PRs, but at the very least we have one more blocker:
* https://github.com/opencontainers/runc/issues/4047 (which I'm actively working one, hope to come to a closure this week)
(and I guess @cyphar can add more)

**Comment by AkihiroSuda** [comment]
> Well, we have the 1.2.0 milestone (which I am trying comb through regularly), and everything that's in there is kind of a blocker for 1.2.0.

Not everything. 
At least this one can be postponed to v1.3 or later
- https://github.com/opencontainers/runc/pull/2523

**Comment by kolyshkin** [comment]
> Not everything.

Well, strictly speaking, everything that is under that milestone needs to either be included into 1.2.0, or be reassigned to a different milestone. Whis is why I said

> Feel free to modify milestones for any issues/PRs

**Comment by cyphar** [comment]
I still think #3985 is needed for 1.2.0. There are several issues with mount propagation (not to mention the serious restriction of idmapped mounts) that are fixed with #3985.

**Comment by AkihiroSuda** [comment]
> > Blockers for GA: Release runtime-spec v1.1.1
> 
> Given that it took well over 3 years to get 1.1.0 finished (incl. half a year from rc1 to GA), is it really necessary (resp. judicious) to block runc 1.2 on a spec-release?

Necessary, because we have been depending on its main branch:
- https://github.com/opencontainers/runc/pull/3993

**Comment by utam0k** [comment]
This task has been done 👍 
> Release runtime-spec v1.1.1 v1.2.0 https://github.com/opencontainers/runtime-spec/pull/1242

**Comment by testinfected** [comment]
Hi guys,

Eagerly waiting for that 1.2 release with fixes on mount options. What do you think is a realistic release date?

Thanks

**Comment by AkihiroSuda** [comment]
I think we are ready to release rc.2, and release GA after verifying compatibility with Docker, containerd, BuildKit, etc.

cc @opencontainers/runc-maintainers

**Comment by AkihiroSuda** [comment]
Turned out that we need to make a decision on this one too
- https://github.com/opencontainers/runc/pull/4283#issuecomment-2156033671

**Comment by h-vetinari** [comment]
#3922 is noted as a blocker here, and a recent comment notes
> @kolyshkin: This is going to be implemented via https://github.com/opencontainers/runtime-spec/pull/1253

Since https://github.com/opencontainers/runtime-spec/commit/701738418b9555d5213337a0991fd0ffd6c37808 is not yet in a runtime-spec release, does that mean you'll need another spec release to unblock runc 1.2?

From the peanut gallery: Given that #3922 is a year old already (and the issue apparently exists back until runc 1.0.2 at least), I'm skeptical that this really has to block runc 1.2? After being 28 months behind the originally intended release date, perhaps it would be reasonable to stop the scope-creep and just focus on fixing any remaining hard regressions and then release?

<img src="https://github.com/opencontainers/runc/assets/33685575/1e66b39f-bc42-4162-ab81-de79cd8d158b" align="left" height="200" width="425" >

**Comment by AkihiroSuda** [comment]
Let me move #3922 to the "probably deferrable to v1.2.1 or v1.3.0" list

**Comment by rata** [comment]
I've tested runc 1.2.0-rc.2 with containerd and opened a PR to fix the issues (some fixes belonged to the cri-tools repo, that are already merged and included in a patch release). All the fixes belong to the containerd repo, nothing to do here in runc: https://github.com/containerd/containerd/pull/10449.

Can someone please take a look at the moby issue? https://github.com/moby/moby/pull/47666

**Comment by kolyshkin** [comment]
Looking into https://github.com/moby/moby/pull/47666, as well as other dmz issues, I'm thinking maybe we should remove it entirely from runc-1.2?

**Comment by rata** [comment]
@kolyshkin are you sure that is dmz related? That moby PR is compiling with nodmz, there might be something on the say we disable it, but it felt like another issue given that it fails in the same way with and without the nodmz buildtag

**Comment by kolyshkin** [comment]
> @kolyshkin are you sure that is dmz related? That moby PR is compiling with nodmz, there might be something on the say we disable it, but it felt like another issue given that it fails in the same way with and without the nodmz buildtag

Yeah, probably not. I will take a look.

**Comment by rata** [comment]
It was indeed something on the way `runc_nodmz` is disabled, work-around here: https://github.com/opencontainers/runc/pull/4345 (still need to fix cc_platform.mk).

I didn't manage to debug the failing tests now that docker compiles and need to leave now. Please someone else take a look :)

To use a patched runc, I used this moby PR to just compile my fork: https://github.com/moby/moby/pull/48161. Compilation works fine here, but the test fail. You can use something like that to try fixing the tests on the CI.

**Comment by kolyshkin** [comment]
I figured out the runc compile issue in moby; see https://github.com/moby/moby/pull/48160

**Comment by apostasie** [comment]
Hey folks.

Would it be possible to cut an RC3? Ideally that would allow closing https://github.com/containerd/nerdctl/pull/3153

Thanks!

**Comment by AkihiroSuda** [comment]
If somebody can take a look at incompatibility issues with Moby (https://github.com/moby/moby/pull/48336), that will be helpful for releasing the next RC (and GA)

**Comment by rata** [comment]
@AkihiroSuda I can't repro the issue locally (even without apparmor=0 in the kernel cmdline), I created a VM in Azure with a stock ubuntu (to see if there was something in that setup), but all test seem to pass just fine. I've tested CI without apparmor, etc, no differences (I'll update the moby PR with more details). @kolyshkin has been testing several things also and he can't repro either. 

The issues in CI seem real to me, its' 100% reproducible that the new runc causes failures and the old doesn't. I guess there is something in the moby CI setup that we don't have in the other envs we are testing this.

@AkihiroSuda Can you have a look, please?
@thaJeztah can you have a look too, pretty please?

**Comment by AkihiroSuda** [comment]
f2f16213e174fb63e931fe0546bbbad1d9bbed6f `init: close internal fds before execve` seems closing stdout too earlier

- https://github.com/moby/moby/pull/48366#issuecomment-2308445022

**Comment by rata** [comment]
Okay, I think I understand now the issues with moby: https://github.com/opencontainers/runc/issues/4384#issuecomment-2318191479. They are on moby, though, and not runc. @thaJeztah it would be great if you can take a look though, as it is quite moby related.

IMHO we can tag a new rc now. @AkihiroSuda @kolyshkin @cyphar @lifubang @thaJeztah (or any other maintainer reading this) what do you think?

**Comment by lifubang** [comment]
> IMHO we can tag a new rc now.

I agree, if there is no begin, I'll open the release PR this weekend.

**Comment by cyphar** [comment]
@lifubang I'll do the release this time.

**Comment by AkihiroSuda** [comment]
- https://github.com/containerd/nerdctl/pull/3153

Tried to run nerdctl tests with runc v1.2.0-rc.3, but something seems hanging on Ubuntu 20.04 (cgroup v1)

**Comment by lifubang** [comment]
#4408

**Comment by rata** [comment]
@lifubang thanks! IMHO, we can cut rc4 when @cyphar fixes that issue

**Comment by cyphar** [comment]
I'll send a patch for it later today. (This will be our first logging output in filepath-securejoin, I guess I'll just use logrus...)

**Comment by cyphar** [comment]
I'm flying home today, but I can prepare the release PR and merge it when I get back home in ~30 hours.

**Comment by cyphar** [comment]
Actually @kolyshkin, do we want #4358 or #4418 for rc4? I'll review them when I get home regardless, but since we have #4367 already I guess we can punt on this until rc5/GA?

**Comment by kannon92** [comment]
Hello. We are very interested in v1.2 for Kubernetes. Can we get an update on this release? It seems most blockers have been resolved so I'm not entirely sure what is blocking this.

**Comment by AkihiroSuda** [comment]
> Actually @kolyshkin, do we want #4358 or #4418 for rc4? I'll review them when I get home regardless, but since we have #4367 already I guess we can punt on this until rc5/GA?

Can be v1.3?

**Comment by h-vetinari** [comment]
> Actually @kolyshkin, do we want #4358 or #4418 for rc4? I'll review them when I get home regardless, but since we have #4367 already I guess we can punt on this until rc5/GA?

Pleeeeeeeease stop the scope creep. Both of these are still in heavy discussion (aside from changing 100s of lines of code). They should not be included after rc3 anymore - the only things still going into 1.2.0 should be regression fixes and unavoidable infra work (like unbreaking CI).

This release started out with the ambition "we should switch to time-based releases, let's aim for 3 months"[^1]. How can it be that _years_ later the urge to "but shouldn't we maybe still include that other thing?" is still unchecked? For some more context of why I'm arguing more forcefully than last time -- over a year ago you [said](https://github.com/opencontainers/runc/pull/3963#issuecomment-1666440966)

[^1]: the 1.3 [milestone](https://github.com/opencontainers/runc/milestone/14) still has its target release date set to Jun 13th 2022; 3 months after the original target date for 1.2

> I don't really see any point to doing a 1.2 release and then doing a 1.3 release a few weeks later, when we can instead get a few remaining PRs in and then do a 1.2 release.

The point would have been that ~1.5 years of feature work until mid-2023 would have become available to users over a year earlier. Please have a read of https://wg21.link/p1000 (it may sound unrelated, but I promise it's applicable), in particular the elaboration after
> There are two basic release target choices: Pick the features, or pick the release time, and whichever you pick means relinquishing control over determining the other. It is not possible to control both at once.

IMO the project should hold itself accountable for the goals it sets - either time-based releases (then the "should we include this?" discussions need to go away), or abandon any pretense of actually wanting to release anything within a certain timeframe, but then define a list of things you're planning to achieve "when it's done", and similarly stick to it until it's released. FWIW, I think time-based releases would be _much_ more beneficial.

**Comment by cyphar** [comment]
@AkihiroSuda I was wondering what @kolyshkin's view on it is, given he worked on the update and knows more than me about the #4426. But yes, I would also be in favour of pushing it to 1.3 (afaics this is a long-standing issue).

**Comment by cyphar** [comment]
I went through the 1.2.0 issues again, and AFAICS the only thing left we actually need is #4452 (which fixes a regression related to the Go `/proc/self/exe` cloning we added for 1.2.0, and is a very easy-to-review one-line patch). I can prepare a 1.2.0 release this week once we get that merged.

**Comment by rata** [comment]
@cyphar SGTM :rocket: !

**Comment by kannon92** [comment]
Thank you for your followup. Really looking forward to the PSI features in runc so looking forward to 1.2!

**Comment by AkihiroSuda** [comment]
:tada:
https://github.com/opencontainers/runc/releases/tag/v1.2.0

**Comment by jorydotcom** [comment]
@opencontainers/runc-maintainers Now that this has been released, would you like to do a blog post? Happy to assist with this and/or social posts!
